### ISSUE TITLE
Interfaces for Environments

### ISSUE THREAD

**Issue by ImIOImI** [opentofu/opentofu#949]
### OpenTofu Version

```shell
❯ tofu -v
OpenTofu v1.6.0-dev
on linux_amd64
```

### Use Cases

In software engineering we have a major problem where we can have two or more things that are related, and need to enforce a common API between them and we don't want to strongly couple them together. We have solved this by creating interfaces. In Terraform, we often have the same problem. We have two or more things, for example, resources for the Kubernetes control plane, node pools, and helm provisioning charts (let's use datadog, external dns, and istio as examples of some helm charts that you might want to capture in IAC). In this example, we have node pools that need to know about the k8s cluster and a collection of helm charts that need to know how to connect to the k8s cluster.

Ideally, we would want to separate all these different configurations into as small units as possible to limit the blast radius of potential updates, decrease code conflicts when multiple engineers are modifying code in parallel, and to decrease the run time of any potential change.

```
─── terraform
    ├── k8s
    │   ├── control-plane
    │   │   ├── main.tf
    │   │   └── outputs.tf
    │   └── node-pools
    │       └── main.tf
    └── helm
        ├── datadog
        │   └── main.tf
        ├── dns
        │   └── main.tf
        └── istio
            └── main.tf
```

### Attempted Solutions

The most common solution I've seen to achieve the above (providing you don't want to lump everything together in a module of modules) will be for each of the dependant modules to include refernces to the kubernetes control plane's state file like:

```
data "terraform_remote_state" "control-plane" {
  backend   = "azurerm"
  workspace = var.context

  config = {
    resource_group_name  = "<resource name>"
    storage_account_name = "<storage account name>"
    container_name       = "<container name>"
    key                  = "<key to remote environment>"
  }
}
```

Then pulling all the needed outputs from the map of outputs. The problem here is that each environment that does this needs to have knowledge of both the state location and the data structure in the outputs map. This makes refactoring a huge pain.

The best solution I've found is to take the above and go a step further and create a module that references the state of an active environment and it surfaces the outputs from that statefile like:

```

output "<output name in remote environment>" {
  value = data.terraform_remote_state.control-plane.outputs.<output name in remote environment>
}

output "kubernetes_cluster_id" {
  value = data.terraform_remote_state.control-plane.outputs.kubernetes_cluster_id
}

output "kubernetes_cluster_name" {
  value = data.terraform_remote_state.control-plane.outputs.kubernetes_cluster_name
}

output "oidc_issuer_url" {
  value = data.terraform_remote_state.control-plane.outputs.oidc_issuer_url
}

...
data "terraform_remote_state" "control-plane" {
  backend   = "azurerm"
  workspace = var.context

  config = {
    resource_group_name  = "<resource name>"
    storage_account_name = "<storage account name>"
    container_name       = "<container name>"
    key                  = "<key to remote environment>"
  }
}
```

Then you can include that module in multiple dependent environments and consume its outputs. 

This requires us to now maintain an environment and a module for each parent environment and keep them in sync (any new changes need to be added to the interface), To that end, I've even created (badly) [a python script that I use in a github action](https://github.com/ImIOImI/github-workflow-build-terraform-interface) that parses an environment and pulls out the output and generates interfaces like the above. 

The benefit here is that there's one file that holds the contract between an environment and its dependencies and we don't have to store all the related environments in a module of modules to facilitate communications about state values between environments. Another great benefit of this type of design is that it can be easily parsed by an IDE and doesn't break autocompletion. 

This is adequate, but largely unnecessary in the majority of simple inheritance cases. It also presupposes that the interface has knowledge about where the state file is for the referenced environment when said environment should be the only source of truth for its state configuration. 

### Proposal

I propose a new block type called an interface (alternate names could be `state` or `environment`, etc.). Its purpose would be to 
1) easily surface outputs in a remote environment WITHOUT running the terraform code within.
2) provide an easy API for interfacing with the remote's environments state file and pulling state information from said environment without the calling environment having to have knowledge about the remote environment's state file. 
3) (hopefully) allow for autocompletion of outputs from remote environments inside calling environments
4) trivialize refactoring of otherwise complex inheritance patterns

It would be called in a very similar manner to a module:

```
interface "k8s" {
  # This should be defined the same way as a module with the exception being there's no need to talk to a module repo
  source= "<path to environment>/<github.com/org/module>"
  # defaults to `default` for the default workspace
  workspace="dev"
}
```

and it could be consumed like:
```
resource "azurerm_federated_identity_credential" "this" {
  name                = "${interface.k8s-data.kubernetes_cluster_name}-ServiceAccount-${local.namespace}-${local.name}"
  resource_group_name = interface.core.resource_group_name
  audience            = ["api://AzureADTokenExchange"]
  issuer              = interface.k8s.oidc_issuer_url
  parent_id           = azurerm_user_assigned_identity.this.id
  subject             = "system:serviceaccount:${local.namespace}:${local.service_account_name}"
}

provider "kubernetes" {
  config_context = interface.k8s.kubernetes_cluster_name
}
```

**Comment by cube2222** [comment]
Hey, thanks for creating the issue @ImIOImI!

My personal (not the core team's) thoughts on this:

It seems like this feature proposal is built around remote state access, which is something that is heavily discouraged overall. The recommended way to share data across statefiles is to use explicit resources (created by the producer) later read by all the consumers, this is also because it avoids consumers having access to all internal sensitive values of the producer stack. More on this in the docs: https://opentofu.org/docs/language/state/remote-state-data

In practice, I believe that this is one of the exact problems solved by various orchestration tools built on top of OpenTofu, both non-commercial ones (like Terragrunt) and commercial ones (all the TACOS).

In other words, I am not supportive of further efforts in directions that rely on remote state access.

**Comment by ImIOImI** [comment]
@cube2222 I totally understand the concern. However, I don't feel like your suggested alternatives are adequate. The idea that I need to pay money for a TACOS or use Terragrunt to make up for some deficiencies in the access model between environments seems like we're actively discouraging people from using an architecture that has a lot of benefits for some orgs.

Two of what I see as generally good alternatives to breaking up environments have their own sets of problems. I can create a module of modules, which can have huge, tedious runs that cross infra engineering domains. I could also use data lookups, but that requires the consuming environment to know things about the producing environment like resource groups, names, regions, etc. that create refactoring nightmares.

It's my experience (and perhaps you don't share this) that smaller environments are better. It tends to create a good separation of concerns. Teams can manage parts of their infra that can be consumed by other teams via published interfaces, blast radius is easier to manage, execution times decrease, and the ability to refactor your work is greatly increased. The question is whether or not we want to continue pushing people away from this type of architecture or forcing them to use gross `remote_state_data` queries in OpenToFu?

What I'm really hearing in your criticism is that there needs to be something closer to a public/private access pattern implemented in OpenToFu that we use in software engineering all the time. The implementation details of a resource should be private to external APIs whereas outputs should be public. I feel like this could go one of two ways:
1) implementing a different remote api call to the state file that programmatically limits the query to outputs (and possibly deprecating the `remote_state_data` resource which seems to have an absurd potential for abuse and we both kind of hate) or
2) caching the outputs somewhere else and reading that. 

I'm not going to act like I have great insight into the backend or anything and have opinions that are insightful on what's easier or better to implement, but my hope is that we can figure out something better than the mess that Terraform left for us.

**Comment by ImIOImI** [comment]
One more of my thoughts that I've had for a while (but I'm attempting to not propose too many world bending changes in one place) is that besides marking outputs as sensitive we should be able to mark outputs as public/private. That way if an environment is called and ran as a module, the calling environment could have private access... but an environment that calls it in the method described above might only have access to public outputs.

**Comment by cube2222** [comment]
The following is my personal opinion:

With the updated design as it is now, I believe it to only be a not-too-big improvement over using resources and datasources with something like secrets manager or s3 and json objects, while the language additions proposed here are quite significant changes.

Additionally, I think this only solves a small part of the wider "orchestration of multiple state files" problem. Even if such interfaces are present, then there's the problem of running the state files in the right order.

In practice, it's worth noting that terraform **is** trying to solve this with their planned "Stacks" feature. The use-case is basically similar to Terragrunt. Plumbing of outputs and inputs and orchestration of execution. Whether we want to solve this orchestration issue in OpenTofu Core, or whether we decide to leave this to external tools like Terragrunt, is something the Steering Committee is planning to discuss and decide on in the not too distant future.

Either way, I think this warrants waiting for more community voices, so I will be attaching the relevant label.

**Comment by ImIOImI** [comment]
My problem with datasources are two fold:
1) you have to hard code in data source names thus complicating refactoring and making it not apparent how changes in a producing environment will affect a remote consuming environment (especially if its being used by another team in another repo)
2) or you can have some sort of global module defining all the data source names that you need and making everything else inherit it. This is way more dry, but then when every environment is dependent on a global configuration module, any changes to said module need to be tested within the entire org... which totally defeats the purpose of decoupling in the first place. 

Storing data as json objects is a recipe for failure. For example, variables in terraform require typing, but a the same json object  can be an object, list or a set. Again, this pushes the responsibility of understanding the data structure and how to convert it to the consuming module which should never be its job.

You correctly point out that this moves the problem of orchestration to either a TACOS, bash scripts, or Terragrunt. The advantage that this has is that a team can publish a public_backend and write an interface to use it, and share said interface across an org. There doesn't have to be a requirement that consuming environments be in the same stack as the producing environment. This makes it easier to build small tight runs that can be more limited in scope while sharing outputs more broadly.

I too am very interested in Terraform's implementation, but it also appears that a significant portion of it will reside in the cloud and not be released to the community.

**Comment by janosdebugs** [comment]
Hey folks, thank you for all your input. We have discussed this in the core team. As this is a fairly complex feature to implement, we would like to prioritize according to community need. If this feature would be useful to you and resolve problems you have with your current code, please add a reaction to this issue.
